File: src/elements.clj
(import* ["./base.js" "./sanctuary.clj" "./prelude.clj"]

(let
   [text
      (lambda [value]
         {:type :text
          :value value})

    replace (lambda [this that text] (.replace this that text))
    canonicalize-children
      (compose (map (when string?
                          (pipe [lines
                                 (fold-map String (replace (regex "" "^[ ]+") " "))
                                 (replace (regex "g" " -- ") "\u2009\u2014\u2009")
                                 text])))
               (unless array? (of Array)))

    block-element
      (lambda [tag-name attrs children]
         {:type :element
          :tag-name tag-name
          :format "block"
          :self-closing false
          :attrs attrs
          :children (canonicalize-children children)})

    inline-element
      (lambda [tag-name attrs children]
         (let [children (canonicalize-children children)]
            {:type :element
             :tag-name tag-name
             :format (if (any (lambda [node] (equals "block" (:format node))) children)
                         "block"
                         "inline")
             :self-closing false
             :attrs attrs
             :children children}))

    self-closing-element
      (lambda [tag-name attrs]
         {:type :element
          :tag-name tag-name
          :format "inline"
          :self-closing true
          :attrs attrs})

    excerpt
      (lambda [children]
         {:type :excerpt
          :children (canonicalize-children children)})]

   {:canonicalize-children canonicalize-children

    :text text
    :excerpt excerpt
    ; The opening and closing tags of each of the following elements
    ; are always rendered on their own lines.
    :article'               (block-element :article)
    :article                (block-element :article {})
    :blockquote'            (block-element :blockquote)
    :blockquote             (block-element :blockquote {})
    :body'                  (block-element :body)
    :body                   (block-element :body {})
    :div                    (block-element :div)
    :dl'                    (block-element :dl)
    :dl                     (block-element :dl {})
    :footer'                (block-element :footer)
    :footer                 (block-element :footer {})
    :head'                  (block-element :head)
    :head                   (block-element :head {})
    :header'                (block-element :header)
    :header                 (block-element :header {})
    :html'                  (block-element :html)
    :html                   (block-element :html {})
    :linearGradient         (block-element :linearGradient)
    :nav'                   (block-element :nav)
    :nav                    (block-element :nav {})
    :object                 (block-element :object)
    :ol'                    (block-element :ol)
    :ol                     (block-element :ol {})
    :svg                    (block-element :svg)
    :ul'                    (block-element :ul)
    :ul                     (block-element :ul {})
    ; The opening and closing tags of each of the following elements
    ; are rendered inline unless the element contains an element
    ; whose opening and closing tags are rendered on their own lines.
    :a'                    (inline-element :a)
    :a      (lambda [href] (inline-element :a {:href href}))
    :aside'                (inline-element :aside)
    :aside                 (inline-element :aside {})
    :code'                 (inline-element :code)
    :code                  (inline-element :code {})
    :dd'                   (inline-element :dd)
    :dd                    (inline-element :dd {})
    :del'                  (inline-element :del)
    :del                   (inline-element :del {})
    :dt'                   (inline-element :dt)
    :dt                    (inline-element :dt {})
    :em'                   (inline-element :em)
    :em                    (inline-element :em {})
    :h1'                   (inline-element :h1)
    :h1                    (inline-element :h1 {})
    :h2'                   (inline-element :h2)
    :h2                    (inline-element :h2 {})
    :h3'                   (inline-element :h3)
    :h3                    (inline-element :h3 {})
    :h4'                   (inline-element :h4)
    :h4                    (inline-element :h4 {})
    :h5'                   (inline-element :h5)
    :h5                    (inline-element :h5 {})
    :h6'                   (inline-element :h6)
    :h6                    (inline-element :h6 {})
    :i'                    (inline-element :i)
    :i                     (inline-element :i {})
    :ins'                  (inline-element :ins)
    :ins                   (inline-element :ins {})
    :li'                   (inline-element :li)
    :li                    (inline-element :li {})
    :p'                    (inline-element :p)
    :p                     (inline-element :p {})
    :pre'                  (inline-element :pre)
    :pre                   (inline-element :pre {})
    :script                (inline-element :script)
    :span                  (inline-element :span)
    :strong'               (inline-element :strong)
    :strong                (inline-element :strong {})
    :time                  (inline-element :time)
    :title'                (inline-element :title)
    :title                 (inline-element :title {})
    ; The following self-closing elements are always rendered inline.
    :hr'                   (self-closing-element :hr)
    :hr                    (self-closing-element :hr {})
    :img                   (self-closing-element :img)
    :link                  (self-closing-element :link)
    :meta                  (self-closing-element :meta)
    :param                 (self-closing-element :param)
    :path                  (self-closing-element :path)
    :stop                  (self-closing-element :stop)}))

File: src/elements.js
'use strict';

module.exports = ($$ => {

  $$[Symbol.for ('text')] = value => ({
    [Symbol.for ('type')]: Symbol.for ('text'),
    [Symbol.for ('value')]: value,
  });

  $$[Symbol.for ('canonicalize-attrs')] = (
    $$[Symbol.for ('reduce-object')]
      (k => v => $$[Symbol.for ('insert')]
                   (k)
                   ($$[Symbol.for ('if-else')]
                      ($$[Symbol.for ('symbol?')])
                      ($$[Symbol.for ('symbol->string')])
                      (String)
                      (v)))
      ({})
  );

  $$[Symbol.for ('replace')] = this_ => that => text => (
    text.replace (this_, that)
  );

  $$[Symbol.for ('canonicalize-children')] = (
    $$[Symbol.for ('compose')]
      ($$[Symbol.for ('map')]
         ($$[Symbol.for ('when')]
            ($$[Symbol.for ('string?')])
            ($$[Symbol.for ('pipe')]
               ([$$[Symbol.for ('lines')],
                 $$[Symbol.for ('fold-map')]
                   (String)
                   ($$[Symbol.for ('replace')]
                      ($$[Symbol.for ('regex')] ('') ('^[ ]+'))
                      (' ')),
                 $$[Symbol.for ('replace')]
                   ($$[Symbol.for ('regex')] ('g') (' -- '))
                   ('\u2009\u2014\u2009'),
                 $$[Symbol.for ('text')]]))))
      ($$[Symbol.for ('unless')]
         ($$[Symbol.for ('array?')])
         ($$[Symbol.for ('of')] (Array)))
  );

  $$[Symbol.for ('block-element')] = tagName => attrs => children => ({
    [Symbol.for ('type')]: Symbol.for ('element'),
    [Symbol.for ('tag-name')]: tagName,
    [Symbol.for ('format')]: Symbol.for ('block'),
    [Symbol.for ('self-closing')]: false,
    [Symbol.for ('attrs')]: $$[Symbol.for ('canonicalize-attrs')] (attrs),
    [Symbol.for ('children')]: $$[Symbol.for ('canonicalize-children')] (children),
  });

  $$[Symbol.for ('inline-element')] = tagName => attrs => _children => {
    const children = $$[Symbol.for ('canonicalize-children')] (_children);
    return {
      [Symbol.for ('type')]: Symbol.for ('element'),
      [Symbol.for ('tag-name')]: tagName,
      [Symbol.for ('format')]: (
        $$[Symbol.for ('any')]
          (node => node[Symbol.for ('format')] === Symbol.for ('block'))
          (children) ?
        Symbol.for ('block') :
        Symbol.for ('inline')
      ),
      [Symbol.for ('self-closing')]: false,
      [Symbol.for ('attrs')]: $$[Symbol.for ('canonicalize-attrs')] (attrs),
      [Symbol.for ('children')]: children,
    };
  };

  $$[Symbol.for ('self-closing-element')] = tagName => attrs => ({
    [Symbol.for ('type')]: Symbol.for ('element'),
    [Symbol.for ('tag-name')]: tagName,
    [Symbol.for ('format')]: Symbol.for ('inline'),
    [Symbol.for ('self-closing')]: true,
    [Symbol.for ('attrs')]: $$[Symbol.for ('canonicalize-attrs')] (attrs),
  });

  $$[Symbol.for ('excerpt')] = children => ({
    [Symbol.for ('type')]: Symbol.for ('excerpt'),
    [Symbol.for ('children')]: $$[Symbol.for ('canonicalize-children')] (children),
  });

  $$[Symbol.for ("html'")] = $$[Symbol.for ('block-element')] (Symbol.for ('html'));

  $$[Symbol.for ("head'")] = $$[Symbol.for ('block-element')] (Symbol.for ('head'));
  $$[Symbol.for ("title'")] = $$[Symbol.for ('inline-element')] (Symbol.for ('title'));
  $$[Symbol.for ('link')] = $$[Symbol.for ('self-closing-element')] (Symbol.for ('link'));
  $$[Symbol.for ('meta')] = $$[Symbol.for ('self-closing-element')] (Symbol.for ('meta'));

  $$[Symbol.for ("body'")] = $$[Symbol.for ('block-element')] (Symbol.for ('body'));
  $$[Symbol.for ("article'")] = $$[Symbol.for ('block-element')] (Symbol.for ('article'));
  $$[Symbol.for ("nav'")] = $$[Symbol.for ('block-element')] (Symbol.for ('nav'));
  $$[Symbol.for ("aside'")] = $$[Symbol.for ('block-element')] (Symbol.for ('aside'));
  $$[Symbol.for ("h1'")] = $$[Symbol.for ('block-element')] (Symbol.for ('h1'));
  $$[Symbol.for ("h2'")] = $$[Symbol.for ('block-element')] (Symbol.for ('h2'));
  $$[Symbol.for ("h3'")] = $$[Symbol.for ('block-element')] (Symbol.for ('h3'));
  $$[Symbol.for ("h4'")] = $$[Symbol.for ('block-element')] (Symbol.for ('h4'));
  $$[Symbol.for ("h5'")] = $$[Symbol.for ('block-element')] (Symbol.for ('h5'));
  $$[Symbol.for ("h6'")] = $$[Symbol.for ('block-element')] (Symbol.for ('h6'));
  $$[Symbol.for ("header'")] = $$[Symbol.for ('block-element')] (Symbol.for ('header'));
  $$[Symbol.for ("footer'")] = $$[Symbol.for ('block-element')] (Symbol.for ('footer'));

  $$[Symbol.for ("p'")] = $$[Symbol.for ('inline-element')] (Symbol.for ('p'));
  $$[Symbol.for ("hr'")] = $$[Symbol.for ('self-closing-element')] (Symbol.for ('hr'));
  $$[Symbol.for ("pre'")] = $$[Symbol.for ('inline-element')] (Symbol.for ('pre'));
  $$[Symbol.for ("blockquote'")] = $$[Symbol.for ('block-element')] (Symbol.for ('blockquote'));
  $$[Symbol.for ("ol'")] = $$[Symbol.for ('block-element')] (Symbol.for ('ol'));
  $$[Symbol.for ("ul'")] = $$[Symbol.for ('block-element')] (Symbol.for ('ul'));
  $$[Symbol.for ("li'")] = $$[Symbol.for ('inline-element')] (Symbol.for ('li'));
  $$[Symbol.for ("dl'")] = $$[Symbol.for ('block-element')] (Symbol.for ('dl'));
  $$[Symbol.for ("dt'")] = $$[Symbol.for ('inline-element')] (Symbol.for ('dt'));
  $$[Symbol.for ("dd'")] = $$[Symbol.for ('inline-element')] (Symbol.for ('dd'));
  $$[Symbol.for ('div')] = $$[Symbol.for ('block-element')] (Symbol.for ('div'));

  return {
    [Symbol.for ('canonicalize-children')]: $$[Symbol.for ('canonicalize-children')],

    [Symbol.for ('text')]: $$[Symbol.for ('text')],
    [Symbol.for ('excerpt')]: $$[Symbol.for ('excerpt')],

    [Symbol.for ("article'")]: $$[Symbol.for ("article'")],
    [Symbol.for ('article')]: $$[Symbol.for ("article'")] ({}),
    [Symbol.for ("blockquote'")]: $$[Symbol.for ("blockquote'")],
    [Symbol.for ('blockquote')]: $$[Symbol.for ("blockquote'")] ({}),
    [Symbol.for ("body'")]: $$[Symbol.for ("body'")],
    [Symbol.for ('body')]: $$[Symbol.for ("body'")] ({}),
    [Symbol.for ('div')]: $$[Symbol.for ('div')],
    [Symbol.for ("dl'")]: $$[Symbol.for ("dl'")],
    [Symbol.for ('dl')]: $$[Symbol.for ("dl'")] ({}),
    [Symbol.for ("footer'")]: $$[Symbol.for ("footer'")],
    [Symbol.for ('footer')]: $$[Symbol.for ("footer'")] ({}),
    [Symbol.for ("head'")]: $$[Symbol.for ("head'")],
    [Symbol.for ('head')]: $$[Symbol.for ("head'")] ({}),
    [Symbol.for ("header'")]: $$[Symbol.for ("header'")],
    [Symbol.for ('header')]: $$[Symbol.for ("header'")] ({}),
    [Symbol.for ("html'")]: $$[Symbol.for ("html'")],
    [Symbol.for ('html')]: $$[Symbol.for ("html'")] ({}),
    [Symbol.for ('linearGradient')]: $$[Symbol.for ('block-element')] (Symbol.for ('linearGradient')),
    [Symbol.for ("nav'")]: $$[Symbol.for ("nav'")],
    [Symbol.for ('nav')]: $$[Symbol.for ("nav'")] ({}),
    [Symbol.for ('object')]: $$[Symbol.for ('block-element')] (Symbol.for ('object')),
    [Symbol.for ("ol'")]: $$[Symbol.for ("ol'")],
    [Symbol.for ('ol')]: $$[Symbol.for ("ol'")] ({}),
    [Symbol.for ('svg')]: $$[Symbol.for ('block-element')] (Symbol.for ('svg')),
    [Symbol.for ("ul'")]: $$[Symbol.for ("ul'")],
    [Symbol.for ('ul')]: $$[Symbol.for ("ul'")] ({}),
    [Symbol.for ("a'")]: $$[Symbol.for ('inline-element')] (Symbol.for ('a')),
    [Symbol.for ('a')]: href => $$[Symbol.for ('inline-element')] (Symbol.for ('a')) ({[Symbol.for ('href')]: href}),
    [Symbol.for ("aside'")]: $$[Symbol.for ("aside'")],
    [Symbol.for ('aside')]: $$[Symbol.for ("aside'")] ({}),
    [Symbol.for ("code'")]: $$[Symbol.for ('inline-element')] (Symbol.for ('code')),
    [Symbol.for ('code')]: $$[Symbol.for ('inline-element')] (Symbol.for ('code')) ({}),
    [Symbol.for ("dd'")]: $$[Symbol.for ("dd'")],
    [Symbol.for ('dd')]: $$[Symbol.for ("dd'")] ({}),
    [Symbol.for ("del'")]: $$[Symbol.for ('inline-element')] (Symbol.for ('del')),
    [Symbol.for ('del')]: $$[Symbol.for ('inline-element')] (Symbol.for ('del')) ({}),
    [Symbol.for ("dt'")]: $$[Symbol.for ("dt'")],
    [Symbol.for ('dt')]: $$[Symbol.for ("dt'")] ({}),
    [Symbol.for ("em'")]: $$[Symbol.for ('inline-element')] (Symbol.for ('em')),
    [Symbol.for ('em')]: $$[Symbol.for ('inline-element')] (Symbol.for ('em')) ({}),
    [Symbol.for ("h1'")]: $$[Symbol.for ("h1'")],
    [Symbol.for ('h1')]: $$[Symbol.for ("h1'")] ({}),
    [Symbol.for ("h2'")]: $$[Symbol.for ("h2'")],
    [Symbol.for ('h2')]: $$[Symbol.for ("h2'")] ({}),
    [Symbol.for ("h3'")]: $$[Symbol.for ("h3'")],
    [Symbol.for ('h3')]: $$[Symbol.for ("h3'")] ({}),
    [Symbol.for ("h4'")]: $$[Symbol.for ("h4'")],
    [Symbol.for ('h4')]: $$[Symbol.for ("h4'")] ({}),
    [Symbol.for ("h5'")]: $$[Symbol.for ("h5'")],
    [Symbol.for ('h5')]: $$[Symbol.for ("h5'")] ({}),
    [Symbol.for ("h6'")]: $$[Symbol.for ("h6'")],
    [Symbol.for ('h6')]: $$[Symbol.for ("h6'")] ({}),
    [Symbol.for ("i'")]: $$[Symbol.for ('inline-element')] (Symbol.for ('i')),
    [Symbol.for ('i')]: $$[Symbol.for ('inline-element')] (Symbol.for ('i')) ({}),
    [Symbol.for ("ins'")]: $$[Symbol.for ('inline-element')] (Symbol.for ('ins')),
    [Symbol.for ('ins')]: $$[Symbol.for ('inline-element')] (Symbol.for ('ins')) ({}),
    [Symbol.for ("li'")]: $$[Symbol.for ("li'")],
    [Symbol.for ('li')]: $$[Symbol.for ("li'")] ({}),
    [Symbol.for ("p'")]: $$[Symbol.for ("p'")],
    [Symbol.for ('p')]: $$[Symbol.for ("p'")] ({}),
    [Symbol.for ("pre'")]: $$[Symbol.for ("pre'")],
    [Symbol.for ('pre')]: $$[Symbol.for ("pre'")] ({}),
    [Symbol.for ('script')]: $$[Symbol.for ('inline-element')] (Symbol.for ('script')),
    [Symbol.for ('span')]: $$[Symbol.for ('inline-element')] (Symbol.for ('span')),
    [Symbol.for ("strong'")]: $$[Symbol.for ('inline-element')] (Symbol.for ('strong')),
    [Symbol.for ('strong')]: $$[Symbol.for ('inline-element')] (Symbol.for ('strong')) ({}),
    [Symbol.for ('time')]: $$[Symbol.for ('inline-element')] (Symbol.for ('time')),
    [Symbol.for ("title'")]: $$[Symbol.for ("title'")],
    [Symbol.for ('title')]: $$[Symbol.for ("title'")] ({}),
    [Symbol.for ("var'")]: $$[Symbol.for ('inline-element')] (Symbol.for ('var')),
    [Symbol.for ('var')]: $$[Symbol.for ('inline-element')] (Symbol.for ('var')) ({}),
    [Symbol.for ('video')]: $$[Symbol.for ('inline-element')] (Symbol.for ('video')),
    [Symbol.for ('embed')]: $$[Symbol.for ('self-closing-element')] (Symbol.for ('embed')),
    [Symbol.for ("hr'")]: $$[Symbol.for ("hr'")],
    [Symbol.for ('hr')]: $$[Symbol.for ("hr'")] ({}),
    [Symbol.for ('img')]: $$[Symbol.for ('self-closing-element')] (Symbol.for ('img')),
    [Symbol.for ('link')]: $$[Symbol.for ('link')],
    [Symbol.for ('meta')]: $$[Symbol.for ('meta')],
    [Symbol.for ('param')]: $$[Symbol.for ('self-closing-element')] (Symbol.for ('param')),
    [Symbol.for ('path')]: $$[Symbol.for ('self-closing-element')] (Symbol.for ('path')),
    [Symbol.for ('stop')]: $$[Symbol.for ('self-closing-element')] (Symbol.for ('stop')),
  };

}) (
  Object.create (
    ['./base.js', './sanctuary.js', './prelude.js'].reduce (
      ($$, path) => Object.assign (Object.create ($$), require (path)),
      Object.create (null),
    ),
  ),
);
